### ISSUE TITLE
Repository "protobuf" has a dirty worktree on upgrade

### ISSUE THREAD

**Issue by rakanalh** [doomemacs/doomemacs#7311]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.

- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.

- [X] The issue can be reproduced on the **latest** available commit of Doom.

- [X] The issue can be reproduced on a stable release of Emacs, such as 27 or 28. *(Doom does not support development builds like 29+ or any version ending in .50 or .9x)*

### Expected behavior

`doom upgrade` without any local modification to the dependencies should not have to ask every time about dirty working tree.

### Current behavior

```
Repository "protobuf" has a dirty worktree:

       M third_party/abseil-cpp

         1) Abort
         2) Stash changes
         3) Discard changes (Choose this if unsure)

      How to proceed? (1, 2, 3; don't know? Pick 3) 3
      Discard all local changes permanently? (y or n) y
    ✓ (32/265) protobuf: d4db41d -> 813fcad [3 commit(s)]
```

That is, every time doom upgrade is executed.

Even though:
```
❯ cd /home/rakan/.emacs.d/.local/straight/repos/protobuf/.git/modules/third_party/abseil-cpp

modules/third_party/abseil-cpp (78be636)
❯ git status
HEAD detached at 78be6368
nothing to commit, working tree clean
```

### Steps to reproduce

run `doom upgrade`

### System Information

https://pastebin.com/ghUpq0Xg

**Comment by xz-dev** [comment]
Here's the log:
```
    ✓ (230/262) copilot.el: 581cadd -> f8283e4 [1 commit(s)]
        f8283e4 fix: Better compatible with jsonrpc (#259)
    > Repository "protobuf" has a dirty worktree:
    
       M third_party/abseil-cpp

         1) Abort
         2) Stash changes
         3) Discard changes (Choose this if unsure)

      How to proceed? (1, 2, 3; don't know? Pick 3) 3
      Discard all local changes permanently? (y or n) y
    > Repository "protobuf" has a dirty worktree:
    
       M third_party/abseil-cpp

         1) Abort
         2) Stash changes
         3) Discard changes (Choose this if unsure)

      How to proceed? (1, 2, 3; don't know? Pick 3) 3
      Discard all local changes permanently? (y or n) y
    ✓ (233/262) protobuf: a02a8b8 -> 976a6eb [15 commit(s)]

```

**Comment by hlissner** [comment]
This is because Straight (our package manager) clones packages recursively, by default, and the protobuf-mode Emacs package comes with [a lot of extra baggage](https://github.com/protocolbuffers/protobuf). To get around this, you'll need to explicitly specify not to do so in its :recipe:

```elisp
;;; add to $DOOMDIR/packages.el
;; This should override the default behavior:
(package! protobuf-mode :recipe (:nonrecursive nil))

;; if that doesn't work, try a more complete recipe:
(package! protobuf-mode :recipe (:host github :repo "protocolbuffers/protobuf" :nonrecursive nil))
```

Since Straight cannot "uninit" a submodule yet, you'll have to manually delete the repo before you can `doom sync`:

```sh
rm -rf ~/.config/emacs/.local/straight/repos/protobuf
```

In any case, I'll close this since it isn't an issue Doom can solve, but I'll look into adding a subtree mirror for protobuf on @doomelpa later ([recorded here](https://github.com/orgs/doomelpa/projects/1)).
